### PR TITLE
perf(trie): use local ThreadPool in Parallel::multiproof 

### PR DESCRIPTION
**Pull request by fgimenez**
Aiming to prevent deadlocks when enabling `ParallelProof::multiproof` ref #13260 it allows us to process proofs in chunks, so that we don't spin up tasks for all the existing proofs at once, potentially causing resource exhaustion and rayon contention.

the goal of this change is just to prevent the deadlock we were experimenting with when enabling parallel multiproof, but I also wanted to verify that splitting the workload in chunks doesn't add a significant overhead, fixing the time calculation for the state root task we have consistently better times compared to the regular calculation (state root task time is usually a bit above half of the parallel calculation time)
```
Dec 17 07:39:37 reth-5 reth-holesky[344736]: 2024-12-17T07:39:37.597759Z  INFO engine::tree: State root task finished block=NumHash { number: 2953384, hash: 0xc80f1ab43ca7cbc67c22edd9e32ae6b039efe06dfccc588e1160906259ff616d } total_time=102.27362ms task_state_root=0x1b50b96acf75949b0ace02fcad63ad56242c94cc93ef51f64d8c87309359d330
Dec 17 07:39:37 reth-5 reth-holesky[344736]: 2024-12-17T07:39:37.664179Z DEBUG engine::tree: Calculated state root root_elapsed=191.658237ms block=NumHash { number: 2953384, hash: 0xc80f1ab43ca7cbc67c22edd9e32ae6b039efe06dfccc588e1160906259ff616d }
```

### CODE REVIEW

File: crates/trie/parallel/src/proof.rs [hunk 1]
```diff
@@ -112,26 +116,47 @@ where
...
+        let storage_root_targets_len = storage_root_targets.len();
+
+        debug!(
+            target: "trie::parallel_state_root",
+            total_targets = storage_root_targets_len,
+            "Starting parallel proof generation"
+        );
 
-        // Pre-calculate storage roots for accounts which were changed.
-        tracker.set_precomputed_storage_roots(storage_root_targets.len() as u64);
```
Comment (shekhirin) [inline crates/trie/parallel/src/proof.rs:117]: why was this removed?

File: crates/trie/parallel/src/proof.rs [hunk 2]
```diff
@@ -112,26 +116,47 @@ where
...
+                );
+
+                let task_start = Instant::now();
                 let result = (|| -> Result<_, ParallelStateRootError> {
+                    let provider_start = Instant::now();
                     let provider_ro = view.provider_ro()?;
+                    trace!(
+                        target: "trie::parallel",
+                        ?hashed_address,
+                        provider_time_ms = provider_start.elapsed().as_millis(),
```
Comment (shekhirin) [inline crates/trie/parallel/src/proof.rs:155]: think we can just do
```suggestion
                        provider_time = ?provider_start.elapsed(),
```
here, and it will be get formatted nicely

File: crates/trie/parallel/src/proof.rs [hunk 3]
```diff
@@ -140,19 +165,43 @@ where
                         DatabaseHashedCursorFactory::new(provider_ro.tx_ref()),
                         &hashed_state_sorted,
                     );
+                    trace!(
+                        target: "trie::parallel",
+                        ?hashed_address,
+                        cursor_time_ms = cursor_start.elapsed().as_millis(),
```
Comment (shekhirin) [inline crates/trie/parallel/src/proof.rs:171]: same

File: crates/trie/parallel/src/proof.rs [hunk 4]
```diff
@@ -140,19 +165,43 @@ where
...
-                    .with_branch_node_hash_masks(self.collect_branch_node_hash_masks)
+                    .with_branch_node_hash_masks(collect_masks)
                     .storage_multiproof(target_slots)
-                    .map_err(|e| ParallelStateRootError::Other(e.to_string()))
+                    .map_err(|e| ParallelStateRootError::Other(e.to_string()));
+
+                    trace!(
+                        target: "trie::parallel",
+                        ?hashed_address,
+                        proof_time_ms = proof_start.elapsed().as_millis(),
```
Comment (shekhirin) [inline crates/trie/parallel/src/proof.rs:189]: same

File: crates/trie/parallel/src/proof.rs [hunk 5]
```diff
@@ -140,19 +165,43 @@ where
...
-                if let Err(err) = tx.send(result) {
-                    error!(target: "trie::parallel", ?hashed_address, err_content = ?err.0,  "Failed to send proof result");
+
+                let task_time = task_start.elapsed();
+                if let Err(e) = tx.send(result) {
+                    error!(
+                        target: "trie::parallel",
+                        ?hashed_address,
+                        error = ?e,
+                        task_time_ms = task_time.as_millis(),
```
Comment (shekhirin) [inline crates/trie/parallel/src/proof.rs:202]: same

**Review by shekhirin — APPROVED** [review]
nits, @rkrasiuk PTAL

File: crates/trie/parallel/src/proof.rs (hunk 1 above)
Comment (fgimenez) [reply crates/trie/parallel/src/proof.rs:117]: gone in one of the rebases for sure, recovered

File: crates/trie/parallel/src/proof.rs (hunk 2 above)
Comment (fgimenez) [reply crates/trie/parallel/src/proof.rs]: sure done

**Review by rkrasiuk — APPROVED** [review]
